### PR TITLE
cut: add whitespace option for separating fields

### PR DESCRIPTION
**Pull request by TechHara**
**TLDR**: "-w" flag in cut is an extension option implemented in [FreeBSD](https://www.freebsd.org/cgi/man.cgi?cut). This option proves to be useful when cutting fields separated by one or more whitespaces. GNU coreutils does not currently implement this option. This PR implements -w option and adds test cases. The performance difference compared to FreeBSD's cut implementation is about 9x.

**Example use case**: "ls" or "ps" returns columns separated by multiple inconsistent whitespaces. We can use -w option to select columns, e.g. "ps | cut -w -f1" to print only pids.

**Implementation**: For best performance, I created a new function (cut_fields_whitespace) dedicated for -w option. Performance of "cut -f" before and after this PR is essentially the same, as show below:

Before
```
$ hyperfine -w3 "./target/release/coreutils cut -f2-4,8 data.tsv"
Benchmark 1: ./target/release/coreutils cut -f2-4,8 data.tsv
  Time (mean ± σ):      1.126 s ±  0.006 s    [User: 1.035 s, System: 0.090 s]
  Range (min … max):    1.118 s …  1.138 s    10 runs

$  hyperfine -w3 "./target/release/coreutils cut -f2-4,8 -d' ' data.tsv"
Benchmark 1: ./target/release/coreutils cut -f2-4,8 -d' ' data.tsv
  Time (mean ± σ):      1.026 s ±  0.003 s    [User: 0.940 s, System: 0.086 s]
  Range (min … max):    1.023 s …  1.031 s    10 runs
```

After
```
$ hyperfine -w3 "./target/release/coreutils cut -f2-4,8 data.tsv"
Benchmark 1: ./target/release/coreutils cut -f2-4,8 data.tsv
  Time (mean ± σ):      1.127 s ±  0.021 s    [User: 1.037 s, System: 0.089 s]
  Range (min … max):    1.110 s …  1.169 s    10 runs

$ hyperfine -w3 "./target/release/coreutils cut -f2-4,8 -d' ' data.tsv"
Benchmark 1: ./target/release/coreutils cut -f2-4,8 -d' ' data.tsv
  Time (mean ± σ):      1.045 s ±  0.026 s    [User: 0.956 s, System: 0.088 s]
  Range (min … max):    1.025 s …  1.103 s    10 runs
```

The performance of "cut -w -f" compared to FreeBSD is about 9x faster:
```
$ hyperfine -w3 "./target/release/coreutils cut -f2-4,8 -w data.tsv" "/usr/bin/cut -f2-4,8 -w data.tsv"
Benchmark 1: ./target/release/coreutils cut -f2-4,8 -w data.tsv
  Time (mean ± σ):      1.242 s ±  0.003 s    [User: 1.153 s, System: 0.088 s]
  Range (min … max):    1.240 s …  1.249 s    10 runs

Benchmark 2: /usr/bin/cut -f2-4,8 -w data.tsv
  Time (mean ± σ):     11.529 s ±  0.089 s    [User: 11.385 s, System: 0.131 s]
  Range (min … max):   11.451 s … 11.731 s    10 runs

Summary
  './target/release/coreutils cut -f2-4,8 -w data.tsv' ran
    9.28 ± 0.08 times faster than '/usr/bin/cut -f2-4,8 -w data.tsv'
```

### CODE REVIEW

File: src/uu/cut/src/cut.rs [hunk 1]
```diff
@@ -256,9 +263,99 @@ fn cut_fields_delimiter<R: Read>(
     Ok(())
 }
 
+#[allow(clippy::cognitive_complexity)]
```
Comment (sylvestre) [inline src/uu/cut/src/cut.rs:266]: what about refactoring it a bit to avoid this?

File: src/uu/cut/src/cut.rs [hunk 2]
```diff
@@ -85,6 +87,10 @@ static LONG_HELP: &str = "
         --delimiter (-d) option. Setting the delimiter is optional.
         If not set, a default delimiter of Tab will be used.
 
+        If the -w option is provided, fields will be separated by any number
```
Comment (sylvestre) [inline src/uu/cut/src/cut.rs:90]: can you please add here that it is freebsd specific ?

File: src/uu/cut/src/cut.rs [hunk 3]
```diff
@@ -256,9 +263,99 @@ fn cut_fields_delimiter<R: Read>(
...
+                out.write_all(line)?;
+                if line[line.len() - 1] != newline_char {
+                    out.write_all(&[newline_char])?;
+                }
+            }
+
+            return Ok(true);
+        }
+
+        for &Range { low, high } in ranges {
```
Comment (sylvestre) [inline src/uu/cut/src/cut.rs:294]: this is lacking comments to explain the various steps

File: src/uu/cut/src/cut.rs [hunk 4]
```diff
@@ -563,6 +672,13 @@ pub fn uu_app() -> Command {
                 .help("specify the delimiter character that separates fields in the input source. Defaults to Tab.")
                 .value_name("DELIM"),
         )
+        .arg(
+            Arg::new(options::WHITESPACE_DELIMITED)
+                .short('w')
+                .help("Use any number of whitespace (Space, Tab) to separate fields in the input source.")
```
Comment (sylvestre) [inline src/uu/cut/src/cut.rs:678]: please also repeat here that it is freebsd specific

File: src/uu/cut/src/cut.rs [hunk 5]
```diff
@@ -508,6 +612,11 @@ pub fn uumain(args: impl uucore::Args) -> UResult<()> {
             {
                 Err("invalid input: The '--delimiter' ('-d') option only usable if printing a sequence of fields".into())
             }
+            Mode::Bytes(_, _) | Mode::Characters(_, _)
+                if matches.get_flag(options::WHITESPACE_DELIMITED) =>
+            {
+                Err("invalid input: The '-w' option only usable if printing a sequence of fields".into())
```
Comment (sylvestre) [inline src/uu/cut/src/cut.rs:618]: please add a test to fall into this case

**Review by sylvestre — COMMENTED** [review]
why not
could you please add a comment in
docs/src/extensions.md
about this difference

**Comment by tertsdiepraam** [comment]
I think they're here: https://github.com/freebsd/freebsd-src/tree/373ffc62c158e52cde86a5b934ab4a51307f9f2e/usr.bin/cut/tests

Though this is not very extensive.

**Comment by TechHara** [comment]
> I think they're here: https://github.com/freebsd/freebsd-src/tree/373ffc62c158e52cde86a5b934ab4a51307f9f2e/usr.bin/cut/tests
> 
> Though this is not very extensive.

and some here too, but not much
https://github.com/freebsd/freebsd-src/tree/373ffc62c158e52cde86a5b934ab4a51307f9f2e/contrib/netbsd-tests/usr.bin/cut

File: src/uu/cut/src/cut.rs [hunk 6]
```diff
@@ -256,9 +264,101 @@ fn cut_fields_delimiter<R: Read>(
...
+
+            for _ in 0..=high - low {
+                if print_delim {
+                    out.write_all(out_delim.as_bytes())?;
+                } else {
+                    print_delim = true;
+                }
+
+                match delim_search.next() {
+                    Some((first, last)) => {
```
Comment (sylvestre) [inline src/uu/cut/src/cut.rs:313]: ditto

**Comment by TechHara** [comment]
Hi @sylvestre, is there anything else that needs to be addressed in the PR?

**Comment by TechHara** [comment]
> GNU testsuite comparison:
> 
> ```
> GNU test failed: tests/tail-2/inotify-dir-recreate. tests/tail-2/inotify-dir-recreate is passing on 'main'. Maybe you have to rebase?
> ```

This is not related to this PR and seems to be failing in other PRs as well

**Comment by TechHara** [comment]
The CICD style/lint failures come from the code that I didn't modify. Not sure why they are failing in this PR but not in other PRs. As for spelling error, should I rename `multispace` to `multi_space` perhaps?

**Comment by sylvestre** [comment]
for multispace, add it to the top of the file in the ignore spell list :)

**Comment by TechHara** [comment]
> for multispace, add it to the top of the file in the ignore spell list :)

done!

File: src/uu/cut/src/whitespace_searcher.rs [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+impl<'a> WhitespaceSearcher<'a> {
+    pub fn new(haystack: &'a [u8]) -> WhitespaceSearcher<'a> {
+        WhitespaceSearcher {
+            haystack,
+            position: 0,
+        }
+    }
+}
+
+impl<'a> Iterator for WhitespaceSearcher<'a> {
```
Comment (tertsdiepraam) [inline src/uu/cut/src/whitespace_searcher.rs:26]: This implementation feels a bit backwards to me. Wouldn't it be easier to give ranges (or maybe even slices) of the fields instead of the whitespace? I.e.

```
first    second    third
^---^    ^----^    ^---^     // give these ranges
     ^--^      ^--^          // instead of these
```

I think that would lead to easier code in `cut_fields_whitespace`.

File: src/uu/cut/src/whitespace_searcher.rs [hunk 8]
```diff
@@ -0,0 +1,99 @@
+// This file is part of the uutils coreutils package.
+//
+// (c) Rolf Morel <rolfmorel@gmail.com>
+//
+// For the full copyright and license information, please view the LICENSE
+// file that was distributed with this source code.
+
+// cSpell:ignore multispace
```
Comment (tertsdiepraam) [inline src/uu/cut/src/whitespace_searcher.rs:8]: Nit: we usually write `spell-checker:ignore` (but the effect is the same)

File: src/uu/cut/src/whitespace_searcher.rs [hunk 9]
```diff
@@ -0,0 +1,99 @@
+// This file is part of the uutils coreutils package.
+//
+// (c) Rolf Morel <rolfmorel@gmail.com>
```
Comment (tertsdiepraam) [inline src/uu/cut/src/whitespace_searcher.rs:3]: You're not Rolf Morel are you? I don't think we need this line.

File: src/uu/cut/src/cut.rs [hunk 10]
```diff
@@ -449,9 +556,13 @@ pub fn uumain(args: impl uucore::Args) -> UResult<()> {
                 };
 
                 let only_delimited = matches.get_flag(options::ONLY_DELIMITED);
+                let whitespace_delimited = matches.get_flag(options::WHITESPACE_DELIMITED);
```
Comment (tertsdiepraam) [inline src/uu/cut/src/cut.rs:559]: I think we can make a bit of an improvement here by making the delimiter a new `enum`:
```rust
enum Delimiter {
    Whitespace,
    String(String),
    None,
}
```

File: src/uu/cut/src/cut.rs [hunk 11]
```diff
@@ -449,9 +556,13 @@ pub fn uumain(args: impl uucore::Args) -> UResult<()> {
...
                 let zero_terminated = matches.get_flag(options::ZERO_TERMINATED);
 
                 match matches.get_one::<String>(options::DELIMITER).map(|s| s.as_str()) {
+                    Some(_) if whitespace_delimited => {
+                            Err("invalid input: Only one of --delimiter (-d) or -w option can be specified".into())
+                        }
```
Comment (tertsdiepraam) [inline src/uu/cut/src/cut.rs:565]: What is FreeBSD's behaviour here? Usually options in coreutils just override each other.

File: src/uu/cut/src/whitespace_searcher.rs (hunk 7 above)
Comment (TechHara) [reply src/uu/cut/src/whitespace_searcher.rs:26]: Yes and no. I can come up with two reasons why I'd prefer iterating over delimiters, rather than fields.

1. `--only-delimited`/`-s` option. We need to know for each line whether there is at least one delimiter or not. With a peekable delimiter iterator, this is straightforward: `if delim_iter.peek().is_none() { ... }`
With the field iterator, on the other hand, this adds a bit of complexity as we need to test  if there are >= 2 field iterations.

2. consistency. The original implementation with a char delimiter also iterates over delimiters, not fields. See `Searcher`, `cut_fields_delimiter`, and `cut_fields`. I tried to follow the original logic as much as possible.

File: src/uu/cut/src/cut.rs (hunk 11 above)
Comment (TechHara) [reply src/uu/cut/src/cut.rs:565]: FreeBSD allows only one of `-w` or `-d` options, not both at the same time.

```sh
$ /usr/bin/cut -d' ' -w -f1 input.tsv
usage: cut -b list [-n] [file ...]
       cut -c list [file ...]
       cut -f list [-s] [-w | -d delim] [file ...]
```

https://github.com/freebsd/freebsd-src/blob/373ffc62c158e52cde86a5b934ab4a51307f9f2e/usr.bin/cut/cut.c#L129

File: src/uu/cut/src/whitespace_searcher.rs (hunk 8 above)
Comment (TechHara) [reply src/uu/cut/src/whitespace_searcher.rs]: done

File: src/uu/cut/src/whitespace_searcher.rs (hunk 9 above)
Comment (TechHara) [reply src/uu/cut/src/whitespace_searcher.rs]: done

File: src/uu/cut/src/cut.rs (hunk 10 above)
Comment (TechHara) [reply src/uu/cut/src/cut.rs:566]: done with two options:
```rust
enum Delimiter {
    Whitespace,
    String(String),
}
```

File: src/uu/cut/src/whitespace_searcher.rs (hunk 7 above)
Comment (tertsdiepraam) [reply src/uu/cut/src/whitespace_searcher.rs:24]: That makes sense!

File: src/uu/cut/src/cut.rs (hunk 11 above)
Comment (tertsdiepraam) [reply src/uu/cut/src/cut.rs:572]: I guess that's in line with the behaviour for the other arguments to `cut` as well. Thanks for checking!

File: src/uu/cut/src/cut.rs [hunk 12]
```diff
@@ -337,6 +353,101 @@ fn cut_fields<R: Read>(reader: R, ranges: &[Range], opts: &FieldOptions) -> URes
...
+#[allow(clippy::cognitive_complexity)]
+fn cut_fields<R: Read>(reader: R, ranges: &[Range], opts: &FieldOptions) -> UResult<()> {
+    let newline_char = if opts.zero_terminated { b'\0' } else { b'\n' };
+    match opts.delimiter {
+        Delimiter::Whitespace => cut_fields_whitespace(
+            reader,
+            ranges,
+            opts.only_delimited,
+            newline_char,
+            match opts.out_delimiter {
```
Comment (tertsdiepraam) [inline src/uu/cut/src/cut.rs:365]: Nit: This could be `opts.out_delimiter.unwrap_or("\t")`

File: src/uu/cut/src/cut.rs [hunk 13]
```diff
@@ -337,6 +353,101 @@ fn cut_fields<R: Read>(reader: R, ranges: &[Range], opts: &FieldOptions) -> URes
     Ok(())
 }
 
+#[allow(clippy::cognitive_complexity)]
```
Comment (tertsdiepraam) [inline src/uu/cut/src/cut.rs:356]: I think clippy is actually making a good point here. I think this function should just delegate to the `cut_fields_whitespace`, `cut_fields_delimiter` and a new function. Ideally some of that could would also be deduplicated but we can worry about that later.

File: src/uu/cut/src/cut.rs [hunk 14]
```diff
@@ -111,8 +118,13 @@ struct Options {
     zero_terminated: bool,
 }
 
+enum Delimiter {
+    Whitespace,
+    String(String), // one char long, String because of UTF8 representation
```
Comment (tertsdiepraam) [inline src/uu/cut/src/cut.rs:123]: I know this comment was here before and we shouldn't fix it in this PR, but this is nonsense right? Looks like somebody got confused between a C `char` (i.e. `u8`) and a Rust char, which is actually a full UTF-8 character.

**Review by tertsdiepraam — COMMENTED** [review]
Few more small improvements and then I think this is ready!

File: src/uu/cut/src/cut.rs (hunk 12 above)
Comment (TechHara) [reply src/uu/cut/src/cut.rs:365]: This does not seem to work, since `opts` is passed as immutable and `unwrap_or` consumes the option.

```
error[E0507]: cannot move out of `opts.out_delimiter` which is behind a shared reference
   --> src/uu/cut/src/cut.rs:365:14
    |
365 |             &opts.out_delimiter.unwrap_or("\t".to_owned())
    |              ^^^^^^^^^^^^^^^^^^ move occurs because `opts.out_delimiter` has type `Option<std::string::String>`, which does not implement the `Copy` trait

For more information about this error, try `rustc --explain E0507`.
error: could not compile `uu_cut` due to previous error
```

I could do clone:
```
&opts.out_delimiter.clone().unwrap_or("\t".to_owned())
```

or pass`opts` as mutable, but that would require more edits from the caller side as well

File: src/uu/cut/src/cut.rs (hunk 13 above)
Comment (TechHara) [reply src/uu/cut/src/cut.rs:356]: I agree. I can do that in the next PR -- don't want to make this PR too convoluted :)

File: src/uu/cut/src/cut.rs (hunk 14 above)
Comment (TechHara) [reply src/uu/cut/src/cut.rs:123]: good catch!

File: src/uu/cut/src/cut.rs (hunk 12 above)
Comment (tertsdiepraam) [reply src/uu/cut/src/cut.rs:365]: There's no mutability involved here. It's just that Rust thinks you need ownershio because you ask for a `String`. `opts.out_delimiter.as_deref().unwrap_or("\t")` fixes that.

Comment (TechHara) [reply src/uu/cut/src/cut.rs]: oh I see. Done

File: src/uu/cut/src/cut.rs (hunk 13 above)
Comment (tertsdiepraam) [reply src/uu/cut/src/cut.rs:356]: Okay, but I don't see the warning from clippy if I run it locally. Did you get the warning or did you copy-paste the allow? And if you did, what version of clippy were you using?

Sidenote: please don't resolve conversations if you're not making changes. You're welcome to disagree of course, but then I'd like to resolve it once I think it's settled.

Comment (TechHara) [reply src/uu/cut/src/cut.rs:356]: I didn't add this line -- it was from the original code. See original code line 259.

Comment (TechHara) [reply src/uu/cut/src/cut.rs:356]: > Sidenote: please don't resolve conversations if you're not making changes. You're welcome to disagree of course, but then I'd like to resolve it once I think it's settled.

Oops sorry about that. Understood.

Comment (TechHara) [reply src/uu/cut/src/cut.rs]: I removed clippy allow, since it does not seem to trigger any warning

Comment (tertsdiepraam) [reply src/uu/cut/src/cut.rs]: Excellent! Thank you!

**Comment by TechHara** [comment]
Hi @tertsdiepraam just wanted to check in to see if anything else needs to be addressed with this PR to be merged.

**Comment by tertsdiepraam** [comment]
Yeah I think it's good enough now as far as adding the feature goes. I do think `cut` is due for a refactor that removes some of the duplicated code though.

**Comment by tertsdiepraam** [comment]
Oops forgot to squash again
